Reset stock for absent products in DFC catalog

mkllnk opened the pull request:
#### :information_source: Funded Feature. Please track ALL ASSOCIATED WORK under the associated tracking code `#11678 DFC Orders`

#### What? Why?

- Closes #13190  <!-- Insert issue number here. -->

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->

During DFC product import, we now reset stock for absent products. The exact conditions are:

* The product belongs to the enterprise we are importing to.
* The product is linked to the catalog we are importing.
* The product does not appear in the catalog anymore.

![DFC Product import preview with product to reset](https://github.com/user-attachments/assets/a077a6ef-8be2-437e-b837-dc5dd71da0b3)

#### Review notes

The pull request includes:

* #13182 

New commits start at: Reset stock for absent products in DFC catalog

#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

1. **Product import**
    - Import from a DFC catalog, with stock or on demand.
    - Remove a product from the DFC catalog.
    - Import the catalog again.
    - The removed product should be reset locally so that it disappears from the shop.
      - If the product was on-demand, it should become stock-controlled
2. **Order Cycle Open - Product sync**
    - Import from a DFC catalog, with stock or on demand.
      - Note: the products must be imported to an enterprise that you are the owner of.
    - Remove the product from the DFC catalog.
    - Add imported product to a new order cycle and schedule it to open at the current time.
    - The removed product should be reset locally so that it disappears from the shop.
      - If the product was on-demand, it should become stock-controlled

#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



## Follow-up
* Create a new issue to apply this logic when syncing at order cycle open (ref: https://github.com/openfoodfoundation/openfoodnetwork/pull/13167)

dacook commented on app/controllers/admin/dfc_product_imports_controller.rb:
```diff
@@ -80,5 +81,23 @@ def list_products(catalog)
         ]
       end
     end
+
```
```suggestion
# Reset stock for any variants that were in the catalog, but have since been removed.
# We still keep the product though and don't delete it because it may come back at a later time and we don't want to lose the connection to previous orders.
```

dacook reviewed (commented):
Great, looks good to me. 
I have to admit it took me a while to get my head around the logic, it's complicated! Well done.

I suggest a comment to explain the intent (mostly copied from your issue description)

Also what do you think about conditionally showing the `reset_products` message only when count.positive?

mkllnk commented:
> I suggest a comment to explain the intent (mostly copied from your issue description)

Thank you, done.

> what do you think about conditionally showing the reset_products message only when count.positive?

Yes, that had crossed my mind and I almost did it yesterday. Now I added that condition.

I also added a preview of the products that will be reset. It's not perfect because I'm not passing on the variant ids of products to reset but that should be okay for now. This uncertainty would also occur during automatic import.

I think I have a new idea about how to go forward with this. During manual import, we could give the choice between resetting stock (default behaviour) or unlinking the variant. Once the semantic link is removed, the automatic import won't reset it anymore. But let's wait for someone to have a problem before we suggest a solution. :wink:

dacook commented on app/views/admin/dfc_product_imports/_absent_variant.html.haml:
```diff
@@ -0,0 +1,8 @@
+%tr
+  %td
+    %label
+      ❌
```
😁 I love how emoji make life easier!

dacook commented on app/views/admin/dfc_product_imports/index.html.haml:
```diff
@@ -32,6 +33,7 @@
                 = link_to(existing_product.id, edit_admin_product_path(existing_product))
               - else
                 = t(".new")
+        = render partial: "absent_variant", collection: @absent_items
```
good idea to use a collection partial

dacook reviewed (approved):
Good idea, that makes things much more transparent. And it's really well explained too 💯

rioug reviewed (approved):
Looks good :+1:

RaggedStaff commented:
Further update following slack conversation... 

I don't think we can just remove a product. If the product is back ordered,  the OFN hub could be holding stock (for a mapped variant product). In that scenario,  we want the hub to be able to sell the rest of the stock,  but not back order more. 

In other scenarios (for non-mapped products),  we need to prevent any further sales...  as the hub won't be holding any stock. 

@mkllnk - Will your suggestion (to just amend the on demand flag) work for both of those scenarios?  Do we need different logic,  depending on whether the product has variant mappings or not?

mkllnk commented:
> Will your suggestion (to just amend the on demand flag) work for both of those scenarios?

Let's play it through.

#### Mapped variants, stock control

The variant becomes unavailable. The stock had been calculated from the wholesale stock but should now be reduced to zero.

#### Mapped variants, on-demand

The distributor has left-over stock from the last order cycle and is happy to sell that until sold out. The solution is to set the variant to stock-controlled but not touch the stock level.

#### Non-mapped products, stock control

The product becomes unavailable. We need to set the stock to zero.

#### Non-mapped products, on-demand

Since there was no mapping, we were able to backorder the exact amount we needed. The stock level should be 0 already. A positive stock level is only achieved if orders were cancelled and the backorder couldn't be adjusted (already complete). In that case we have left-over stock and the distributor may be happy to sell that?

### General solution

When we have a product locally that is missing in the remote catalog, we need to distinguish between two cases:

* Stock-controlled: Our stock level is a cache of the remote stock level. Since the product vanished, we need to set the stock level to zero.
* On-demand: We track local stock and our local stock level is independent of the remote catalog. But we can't order any more. So we need to change the product from on-demand to stock-controlled (`backorderable = false`) and don't touch the stock level.

RaggedStaff commented:
> #### Mapped variants, stock control
> 
> The variant becomes unavailable. The stock had been calculated from the wholesale stock but should now be reduced to zero.

Not sure I agree with this case. If it's a mapped variant,  I'm ordering cases,  so the OFN stock reflects locally held stock right? 😕 Does it make a difference whether it's limited or not? Or is OFN tracking when to reorder wholesale variants somewhere else? 

> #### Mapped variants, on-demand

✔️

> #### Non-mapped products, stock control

✔️

> #### Non-mapped products, on-demand

✔️
I think I agree on those 3.

> ### General solution
> 
> When we have a product locally that is missing in the remote catalog, we need to distinguish between two cases:
> 
>     * Stock-controlled: Our stock level is a cache of the remote stock level. Since the product vanished, we need to set the stock level to zero.
> 
>     * On-demand: We track local stock and our local stock level is independent of the remote catalog. But we can't order any more. So we need to change the product from on-demand to stock-controlled (`backorderable = false`) and don't touch the stock level.

I think those are the 2 scenarios,  I'm not sure we're aligned on when to apply them. I think we need to determine when stock reflects locally held stock (for split wholesale cases) and leave that in place (but remove the on demand flag to prevent more back orders).

mkllnk commented:
> > *Mapped variants, stock control*
> 
> > The variant becomes unavailable. The stock had been calculated from the wholesale stock but should now be reduced to zero.
> 
> Not sure I agree with this case. If it's a mapped variant, I'm ordering cases, so the OFN stock reflects locally held stock right? 😕 Does it make a difference whether it's limited or not? Or is OFN tracking when to reorder wholesale variants somewhere else?

That's not how it works at the moment. If the variant in Shopify has limited stock then we just _cache_ the stock level in OFN. We don't remember any local stock. We only remember local stock when the variant is set to on-demand, unlimited stock.

We don't have a way to track the remote stock level and remember left-overs from the last order cycle. We would need two stock levels for that while OFN has only one per variant.

This discussion is now getting out of the scope of this issue about resetting stock. I think it would be a bigger piece of work to introduce two stock levels.

RaggedStaff commented:
> That's not how it works at the moment. If the variant in Shopify has limited stock then we just _cache_ the stock level in OFN. We don't remember any local stock. We only remember local stock when the variant is set to on-demand, unlimited stock.

Ok. So if I have a Product with mapped variants & variants have limited stock, OFN doesn't set it to On Demand ? 😕 I'm really confused - how does OFN handle mapped variant stock levels, if they're not unlimited ? Or are we not handling that yet ? (I checked & Hod's don't have any on the FDC export atm 😅 )

> We don't have a way to track the remote stock level and remember left-overs from the last order cycle. We would need two stock levels for that while OFN has only one per variant.

So, we can't track the remote stock level, but if it's unlimited we are tracking local stock levels, is that correct ? 

> This discussion is now getting out of the scope of this issue about resetting stock. I think it would be a bigger piece of work to introduce two stock levels.

Agreed, and I hope we don't need to.

mkllnk commented:
> how does OFN handle mapped variant stock levels, if they're not unlimited ? Or are we not handling that yet ? 

We don't handle that yet.

mkllnk commented:
> if it's unlimited we are tracking local stock levels, is that correct ?

Correct.

RaggedStaff commented:
Thanks @mkllnk.

I think this makes sense now & it's the most sensible course for now.

1 final follow up question - if OFN was to handle limited stock of mapped variants, how would you go about it (understanding that will be a separate issue) ?

dacook reviewed (approved):
Great explanation 👍

mkllnk commented:
Note to self: need to add reset to automatic product sync that has been merged now.

mkllnk commented:
@dacook, new commits begin at 
* [Remember local stock of backordered products](https://github.com/openfoodfoundation/openfoodnetwork/pull/13191/commits/e35a29cc29700b6db681a8f7b95d34c9349b71a1)

rioug reviewed (approved):
Nice one !

dacook reviewed (approved):
Well done! There's such a world of variables when integrating with a remote system, but it feels like we're keeping on top if it

dacook commented:
@mkllnk because this one requires changing the DFC catalog, does it need to be tested by Garethe? 

Also I've expanded on the testing notes to cover the new cases. Can you please review and update if necessary?

mkllnk commented:
> because this one requires changing the DFC catalog, does it need to be tested by Garethe?

No, I think that Filipe has access to a test store as well.

filipefurtad0 commented:
Hey @mkllnk ,

> Remove a product from the DFC catalog.

I'd need some help on editing the DFC catalog. I've posted [here](https://openfoodnetwork.slack.com/archives/C02TZ6X00/p1743505920349559?thread_ts=1743110720.111819&cid=C02TZ6X00) as well. Added the feedback needed label.

filipefurtad0 commented:
Garethe provided feedback via Slack, I should be able to proceed now :+1:

filipefurtad0 commented:
Hey @mkllnk @dacook @RaggedStaff ,

On the proposed test cases:

**Product Import**

i) If a product is set to on demand (with on_hand = 0) and another one is stock controlled (on_hand = 2), then:

![image](https://github.com/user-attachments/assets/ecbd6f3b-8da1-4a0a-8b28-8af4a9b6457f)

Removing these two products from the catalog on Shopify, and re-importing the catalog:

![image](https://github.com/user-attachments/assets/233b3478-e386-44a8-96d9-2a06d78920e4)

![image](https://github.com/user-attachments/assets/608a0838-f1ca-4a89-a25f-c901b2a4ad79)

![image](https://github.com/user-attachments/assets/c14e6066-9572-40c5-ba48-76b0e7dba5cd)

This removes the products from the shopfront, in both cases, the available stock is zero.

ii) If a product is set to on demand (but has available stock which is non-zero, in the example below, on_hand = 222) and another one is stock controlled (on_hand = 2):

![image](https://github.com/user-attachments/assets/660a28ee-ded5-494c-b113-d25cb4004568)

then, repeating the steps above (removing the products from the DFC catalog, and re-importing them):

![image](https://github.com/user-attachments/assets/ba7392c4-afcd-4c84-aff1-110a89ae84b3)

- for the product set to on demand, but with on_hand = 222 -> the product gets stock controlled, and since it has non-zero available stock, it is not removed from the shopfront. It becomes stock controlled.
- for the product set to on_hand = 2 -> it is removed from the shopfront, as the on_hand is now zero.

**Order Cycle Opening (Product Sync)**

The outcome is the same, on product sync, by opening of an order cycle, i.e.,:

- for the product set to on demand, but with on_hand = 222 -> the product gets stock controlled, and since it has non-zero available stock, it is not removed from the shopfront. It becomes stock controlled.
- for the product set to on_hand = 2 -> it is removed from the shopfront, as the on_hand is now zero.

So, this looks good to merge?

I've noticed a minor regression though, but it is only affecting the UI, and not the import functionality itself:

I've noticed that on product import, the number of successfully imported products is not displayed, as it used to.

After this PR, importing 10 products, displays:

![image](https://github.com/user-attachments/assets/96c6dba5-aa04-463c-b69b-81892d3e3d69)

Before this PR:

![image](https://github.com/user-attachments/assets/ec33c3fd-91ca-42ad-9a48-68a6fcde51b4)

dacook commented:
Great news, thanks Filipe. I'll have a look at that regression tomorrow.

dacook commented:
It took me an embarrassingly long time to figure out what was happening there. I discovered that the count variable has been removed in the translations. I'm not sure why, will ask in #translations

dacook commented on app/views/admin/dfc_product_imports/import.html.haml:
```diff
@@ -3,5 +3,6 @@
 
-%p= t(".imported_products")
-= @count
+%p= t(".imported_products", count: @count)
```
As this is an existing translation, the old translation (without count variable) is already present in about 11 locale files. This results in the `@count` variable no longer printing:

> I've noticed that on product import, the number of successfully imported products is not displayed, as it used to.

https://github.com/openfoodfoundation/openfoodnetwork/pull/13191#issuecomment-2773926148

Ideally, I think we would announce to translators to re-translate (maybe through transifex if there's a feature for that). Or create a different key.
But given that it's going to display exactly as it was before, I will go ahead and update the locale files to resolve this issue.

filipefurtad0 replied:
Thinking about this now, I'm not sure why the spec `./spec/system/admin/dfc_product_import_spec.rb:20` did not catch this regression... It fails on my local system (before you've fixed it @dacook ).

```
  imports from given catalog (FAILED - 1)

Failures:

  1) DFC Product Import imports from given catalog
     Failure/Error: expect(page).to have_content "Imported products: 1"
       expected to find text "Imported products: 1" in "Logged in as : aliza@turner.biz Account Logout Store Dashboard Products Reports Enterprises Products Inventory Import DFC product catalog import Imported products:"
```

sigmundpetersen replied:
Could it be due to different locales between fr_staging and the CI test environment?

rioug commented:
> Gaetan are you ok with the final commit? I _think_ it should be ok but it would be good to have another sense check.

Yeah it should be good.

filipefurtad0 commented:
Hey @dacook ,

I see your latest commit only changes the locales, so I did not perform additional tests on the functionality introduced on this PR, just on how the import page is rendered.

I think changing the `en.yml` file would have triggered change requests on Transifex, so I think this would have been an option, instead of changing all locales yourself (right? Am I missing something?). But I've tested two of them on staging-FR, and I think we're good:

- `en_FR.yml`

![image](https://github.com/user-attachments/assets/6acd4d27-3e24-4f9a-8fcb-53d76633c9cd)

- `fr.yml`

![image](https://github.com/user-attachments/assets/990bb2cd-a040-4fc9-ae09-ddf0782e9d03)

Merging, thanks for addressing this.

dacook commented:
> I think changing the en.yml file would have triggered change requests on Transifex, so I think this would have been an option, instead of changing all locales yourself (right? Am I missing something?).

I'm not completely familiar with Transifex, so I'm interested to learn more! When you say "change requests", do you mean it would appear as "to review" here? As we can see there is a very large number of strings to review, so I suspect it would not happen quickly (if ever).
![Screenshot 2025-04-22 at 10 12 31 am](https://github.com/user-attachments/assets/b4798cfc-cecc-4cdb-88b4-ad8a43caae38)

In this case, it was more of a functional change than a translation change, so I felt it was safe enough to make the change myself, and it will avoid future confusion.
